### PR TITLE
Feature/capture unit test

### PR DESCRIPTION
**Pull request by kaizhangNV**
Convert existing examples into tests for capture/replay layer.

- Add a simply TestBase class to the example base class, which handle the command line input parameter "--test-mode", so we can turn the examples into test mode.
- In the test mode, we first disable the swapchain and enable offline rendering, because we will run those tests on headless environment. Then we enable printing the hashcode of all the entry point functions to stdout.
- Since getting hash code is a slang API, it will recored by our recorder. Therefore, it can be replayed by the replayer, and we will also print the hash code the the stdout.
- The test script will parse the hash code from the stdout, and compare the hash code generated by tests with one generated by replayer.

Fix some bugs in record/replay:

- remove release() call for those com ptr.
- don't store the hash code because it's just a blob data not a handle.

### CODE REVIEW

**Comment by kaizhangNV** [comment]
The changes in `source/slang-record-replay/` directory is just minor bug fix, so feel free to skip it if you are not interested.

File: examples/example-base/example-base.cpp [hunk 1]
```diff
@@ -102,23 +110,27 @@ Slang::Result WindowedAppBase::initializeBase(
...
-    if (frameBufferIndex == -1)
-        return;
 
     gTransientHeaps[frameBufferIndex]->synchronizeAndReset();
     renderFrame(frameBufferIndex);
     gTransientHeaps[frameBufferIndex]->finish();
 }
 
-void WindowedAppBase::createSwapchainFramebuffers()
+void WindowedAppBase::OfflineRender()
```
Comment (csyonghe) [inline examples/example-base/example-base.cpp:119]: offlineRender

File: examples/example-base/test-base.cpp [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        if (wcscmp(szArglist[i], L"--test-mode") == 0)
+        {
+            m_isTestMode = true;
+        }
+    }
+
+    return 0;
+}
+#else
+int TestBase::parseOption(int argc, char** argv)
```
Comment (csyonghe) [inline examples/example-base/test-base.cpp:24]: I guess it will be fine if we just use these version.

File: test-record-replay.sh [hunk 3]
```diff
@@ -0,0 +1,147 @@
+#!/usr/bin/env bash
+
```
Comment (csyonghe) [inline test-record-replay.sh:2]: We cannot run this test on windows without bash. It is best to expose this as a slang-unit-test that uses the Process class to create processes.

File: examples/example-base/example-base.cpp (hunk 1 above)
Comment (kaizhangNV) [reply examples/example-base/example-base.cpp:119]: ack

File: examples/example-base/test-base.cpp (hunk 2 above)
Comment (kaizhangNV) [reply examples/example-base/test-base.cpp:24]: act

**Comment by kaizhangNV** [comment]
Finish part of issue #4570.

**Pull request merged**